### ISSUE TITLE
PGMcrys / list of improvements possible (add more ideas over time):

### ISSUE THREAD

**Issue by E471r** [E471r/FEcrys#5]
1.	**Installation.** Current installation requires a custom conda environment equipped with an outdated set of versions of most of the packages. This is a limitation for expanding the library to use newer methods that may depend on newer versions of certain packages. TODO: update the installation instructions to a newer TensorFlow version.

2.	**Spline initialisation.** Initial spline parameters are currently random, which may increase the time needed to converge a training run at the start of training. TODO: Implement identity initialisation.

3.	**Base distribution.** Since the splines are sufficiently expressive (even with 5 bins), everything is currently fixed to using the uniform base distribution (1/(L**n_DOFs) ; L = spline interval width). A base distribution is completely characterised by two methods (sample_base_ and ln_base_), where ln_base_ provides the absolute log likelihood of the input, and sample_base_ ergodically samples this likelihood in one shot (uncorrelated samples). This is located inside SingleComponent_map* (the representation layer for PGMcrys*), calling this class IC_MAP in this paragraph. TODO: However, any higher temperature distribution can be used instead of uniform (extreme temperature). In the IC_MAP class these methods can be changed to base_dist.sample_base_ and base_dist.sample_base_, respectively. Either methods are not used until the instance(s) of the IC_MAP gets bound to the model (i.e., the trainable spline layers). For this reason it would be possible to swap the base_distribution instance in an IC_MAP instance as soon as the IC_MAP is initialised on the MD data. For example: this would allow initialising ic_map = IC_MAP(MD_data) on the MD data as usual, mapping that data to the representation space x = ic_map.forward(MD_data) and then initialising a custom/user-defined base_dist = BASE_DIST(x) that can replace the default base_dist inside the ic_map. A user-defined BASE_DIST would only need to provide the methods sample_base_ and ln_base_, with anything else left up to the user. Note: The base distribution should be at a higher temperature (wider in all directions) compared to the MD data distribution (in the x space). Taken together, points 2 and 3 would allow faster convergence, that can be almost instant for the 1D marginal distributions, with only correlations remaining for the splines to learn (unless some corrections are also modelled inside the BASE_DIST).

4.	**Energy-based training.** All current training is fixed to maximising the average log likelihood of the model on the MD data, which does not involve force field evaluations during training. TODO: Methods of training that evaluate the MD potential energy surface (force field) cab be added. For example: Equipped with a base distribution that has the same topology as the MD data (i.e., the number islands of finite probability), maximising the likelihood of the model on the samples from the BASE_DIST is very common in the literature. If the number of modes is unknown, the energy-based training can take the form introduced in arXiv:2208.01893. In any case, the forces from the force field would be needed in the back-propagation. Setting this up in 'graph-mode' (e.g., tf.function over the backprop auto-differentiation as currently) would require a custom setup. However, this could still be ran without a special setup in 'eager mode' at a slower training speed; suitable for testing methods.

5.	**Flexibility** The structure of the NN folder could be improved to allow flexibility in benchmarking new hyperparameters, architectures, and training methods.

**Comment by mrshirts** [comment]
@E471r  this is great!

> Installation. Current installation requires a custom conda environment equipped with an outdated set of versions of most of the packages. This is a limitation for expanding the library to use newer methods that may depend on newer versions of certain packages. TODO: update the installation instructions to a newer TensorFlow version.

I think @aiwe2320 has addressed this in https://github.com/shirtsgroup/crystal_mapping, he can talk about what is working. 

> Spline initialisation. Initial spline parameters are currently random, which may increase the time needed to converge a training run at the start of training. TODO: Implement identity initialisation.

@aiwe2320 also addressed this in https://github.com/shirtsgroup/crystal_mapping, he can talk about what is working. 

> Base distribution.

@aiwe2320 has also looked at this, and has created some additional functionality, but it's not clear we are doing things the best way. 

> Energy-based training. 

We are very interested in doing this, and could definitely use help in making this happen.  As you point out, there's a number of things that need to be done.

> Flexibility The structure of the NN folder could be improved to allow flexibility in benchmarking new hyperparameters, architectures, and training methods.

Definitely we are interested in this.  We tried reducing aug_dimension from 25 to 6, reducing the number of parameters from 6278892 to 2380, and it did NOT noticeably hurt the learning curve.

**Comment by aiwe2320** [comment]
Yes, these are very helpful notes, thank you @E471r ! Here are some notes about what we've done with the LJ system toward these goals:

**1) Installation**
I set up a conda environment with TensorFlow 2.19.1 and did some tweaking to the FEcrys code so it would work. The issues I encountered while establishing the updated environment mostly had to do with model saving and loading with pickle. I used AI to resolve this by serializing and deserializing model variables, but the solution is incomplete. It works for our custom representation layers now, but runs into errors in serializing `WhitenFlow` objects that I have not yet resolved.
https://github.com/shirtsgroup/crystal_mapping/blob/main/src/fecrys_util/NN/model_helper.py
https://github.com/shirtsgroup/crystal_mapping/blob/main/defaults/cmp_gpu.yml

**2) Spline initialisation to the identity**
I was able to set up initialisation to the identity (also with the help of AI). This can be done by properly initialising the output layer of the spline such that a) all width and height parameters are equal and b) all slope parameters are one. The former can be achieved by initialising output weights to zero, and the latter can be achieved by specifying the biases such that the softplus output for the spline slopes are all one. I can write up and share the math for this if desired. We have confirmed that this approach initialises to the identity and have a working implementation set up at https://github.com/shirtsgroup/crystal_mapping/blob/main/src/fecrys_util/NN/spline_layer.py

**3) Base distribution**
We've set up different base distributions in the code, but in an ad hoc manner; your suggestion would better for generalization. **Why should the base distribution be wider than the MD distribution in all dimensions?** Would you expect the spline to better capture an expansion compared to a contraction in the data?

**4) Energy based training and architecture flexibility** 
As @mrshirts mentioned, we are interested in these things and hopefully will be able to experiment soon in these areas with the Lennard-Jones system.

**Comment by mrshirts** [comment]
> We've set up different base distributions in the code, but in an ad hoc manner; your suggestion would better for generalization. Why should the base distribution be wider than the MD distribution in all dimensions? Would you expect the spline to better capture an expansion compared to a contraction in the data?

More precisely, whenever you are comparing probability distributions, the probabilities distributions you DO have samples from should be wider than the probability distributions you DON'T have samples from.  This is because you want to make sure that all proposed new configurations have some nonzero probability in the sample space you are generating.    Otherwise, you cannot effectively distinguish the differences between different new configurations you are trying to evaluate. 

But this is comparing the two distributions AFTER any mapping, when you are actually computing log likelihoods.

**Comment by E471r** [comment]
@aiwe2320 and @mrshirts: Thank you for working on this.

@aiwe2320: Training by contraction vs. expansion; In my experience, the contraction of an initially-wide learned distribution (q) around a finite number of training examples is the most well-tested training mechanism, compared to expansion of an initially-narrow q around the same set of training example. The space between the training examples, where the validation data is located, is the only space that is used for FE calculations. This space is very important. We want q(validation data)>0 on this space, and as high as possible on average (proportional to low validation error), because this ensures an unbiased overlap between q and the physical Boltzmann distribution (p(validation data)>0; since the validation data was sampled by MD). Using the uniform base distribution p_{0} (very wide) ensures that q is also very wide at the start. When q is wide-enough in all directions at the start, it is expected to take a similar number of training steps to contract by a certain amount around the training examples from each direction, eventually leading to q overfitting on the training examples by some amount (validation error and standard errors become higher). I think it is best to leave enough space for q to contract by itself during training, because the contraction is quite stochastic; we don’t know which part of the distribution will overfit first (depends on learning rare, initial parameters, the order and number of distinct training examples appearing in the training batches). It would be interesting to check how well an overfit model, when provided with new training data, can expand itself back to a good overlap with the validation data. I remember in our first paper, in the first figure (3D example), the three models were trained like this; overfit first and then swapped the training and validation sets to overfit again. Both halves of the training run behaved similarly, indicating that q can expand and relocate (post-training a model is possible). However, I haven't tried this in higher dimensions.

**Comment by E471r** [comment]
Adding one more TODO: The main and perhaps the most challenging TODO with designing configurational maps is removing the 'last assumption': q(r) = q(P(r)), where q is the model, r is any configuration of the system, and P is a function permutating all of the atoms, or a subset of the atoms, of the same atom-type. So far this was addressed by limiting the representation domain as much as possible to a fixed permutation (i.e., P(x) \ne Identity(x) would be outside of the model domain, similar to how r+v3 for any 3D vector v3 \ne [0,0,0] is outside of the domain since the system is translationally fixed). In flexible molecules removing all the Ps is not always possible or practical. In flexible molecules, the current training of q requires equally ergodic sampling of x and P(x) in the training data, and then assuming that the model learns the invariance relation after enough training (the assumption). To shrink the representation domain (fewer Ps), or to let some Ps create a more ergodic datasets (randomising data on specific orbits), sym.py file was started and actively used (it was needed for correct FEs in many cases). Once q(x) = q(P(x)) is made true on any x by construction, the sym.py file would become redundant. The maps inside q would be a lot more general also. Currently all the maps are used only on curated data (e.g., at a known fixed temperature, with MD data checked carefully, or processed with sum.py, before using it).

TODO related to testing models with non-deterministic likelihoods (beyond normalising flows). If the likelihood q(r) on physical configuration r is stochastic (q(r) \ne q(r) evaluated on two separate occasions), there might be a way to propagate the error on q into the error on f (e.g., f being the FE estimate from MBAR). I think this is the only step needed to start benchmarking augmented normalising flows, and diffusion-based models, recently appearing in the literature. This pathway creates more options for addressing the q(r) = q(P(r)) objective.

**Comment by mrshirts** [comment]
@E471r I agree that treating permutation symmetries as well as possible and be able to handle augmented normalizing flows are definitely useful improvements that are important.  Can you split these topics into more specific lists of todos?

**Comment by E471r** [comment]
@mrshirts About the permutations, I think my paragraph above is slightly incorrect. It would be useful to assert one important point: Train the current model with non-ergodic MD data, but reweight it using completely ergodic validation data. The FEs should be still accurate (always ?) because we never expect the model to be perfect, including how well the permutations are learned. If this is always true, with both BAR and MBAR, the ergodicity assumption is only about the validation data (i.e., the MD data that is not used for training). The step of checking and using sym.py (when needed) on each validation set is currently supervised. So far, I was using this on all MD data, before the training : validation split. The at this step supervision can be reduced by devising an algorithm that applies the sym.py methods to the MD data automatically when needed (can be straightforward to add), to by considering symmetry aware architectures (pure further work). The first step in the later case might be to test augmented normalising flow. This approach learns q(r, _a_) as the learned distribution with the exact and analytical likelihood (no error on q(r, _a_)), but contains the unphysical variable _a_. The step of marginalising q(r) from q(r, _a_) on a finite K number of points drawn from a~p(_a_|r) introduces uncertainty on q(r). [citations missing] This uncertainty depends on K, the choice of p(_a_|r), and how well q(r,_a_) was fitted. Propagating this uncertainty forward into the two-state BAR calculation it expected to help using augmented normalising flows in an unsupervised way. Similar equations with MBAR are also needed. I would start with a toy example where both r and _a_ are one-dimensional before testing on supercells. I am sorry that I cannot contribute in this exciting further work at the moment. All of the main further-work points related to the FEcrys/O/NN folder, paraphrased from my thesis, are now mentioned in this chat.

**Comment by msalvalaglio** [comment]
@aiwe2320, @mrshirts, @E471r - I have forked the repo to leave this as a snapshot of the work described in Edgar's last paper,  and move the joint development to a shared repository. 

I think we should also split the scientific discussion on permutational invariance and ergodicity of the training data from the list of code-related issues.

**Comment by mrshirts** [comment]
> I think we should also split the scientific discussion on permutational invariance and ergodicity of the training data from the list of code-related issues.

Yes, agreed.  I am working on the math of the q(r, a)  vs. r and augmented normalizing flows (which I think is likely to help a lot to improve optimization), and I think that is separate, so there is no code to do this yet.

I admit I am still not sure what sym.py is doing after edgar's description - is it just generating permutated trajectories?

**Comment by E471r** [comment]
I agree the issues are various and some issues are not directly related to the current code. I would approximately score the issues of the current code (5: no issues; 0: big issue) as: ease of using the code by newcomers (3), reproducibility of the PGMcrys paper (5), reproducibility of the published method in supercells with up to 32 molecules using the current code (5 if using sym.py when needed; 4 otherwise), code understandability and documentation (4 with thesis; 3 with papers; 1 otherwise), cost-effectiveness of running the method on supercells with more than 32 molecules (3? without self-attention blocks; 0 with self-attention), ease of customising the code and hyperparameters (3 if documentation is clear; 1 otherwise), generality of the current code (NN folder) to other types of atomistic systems (0). If I started a PhD on this topic again I would start from criticisms and further work discussions centred on the lowest-scoring issues. However, the time and effort needed to turn each above point to 5/5 (fixed) would be difficult to predict from the current view (ultimately requires reproducing similar error bars with the most general bijector and in larger supercells). The reason sym.py is confusing is linked to the poor documentation, especially without the thesis. Yes, sym.py only permutes atoms, to make the MD data more ergodic before training and reweighting. The permutations are specific to be helpful for this current model; atoms are not jumping. A notebook example showing this supervised step is still missing. I believe it is possible to make that run in an unsupervised way.
